File: service/src/test/java/org/coner/core/util/ApiEntityTestUtils.java
package org.coner.core.util;

import java.math.BigDecimal;
import java.time.Year;
import java.util.Date;
import java.util.Set;

import org.coner.core.api.entity.CarApiEntity;
import org.coner.core.api.entity.CompetitionGroupApiEntity;
import org.coner.core.api.entity.CompetitionGroupSetApiEntity;
import org.coner.core.api.entity.EventApiEntity;
import org.coner.core.api.entity.HandicapGroupApiEntity;
import org.coner.core.api.entity.HandicapGroupSetApiEntity;
import org.coner.core.api.entity.PersonApiEntity;
import org.coner.core.api.entity.RegistrationApiEntity;
import org.coner.core.api.entity.RunApiEntity;
import org.coner.core.api.request.AddCompetitionGroupRequest;

import com.google.common.collect.Sets;

public final class ApiEntityTestUtils {

    private ApiEntityTestUtils() {
    }

    public static PersonApiEntity fullPerson() {
        return fullPerson(
                TestConstants.PERSON_ID,
                TestConstants.PERSON_FIRST_NAME,
                TestConstants.PERSON_MIDDLE_NAME,
                TestConstants.PERSON_LAST_NAME
        );
    }

    public static PersonApiEntity fullPerson(String id, String firstName, String middleName, String lastName) {
        PersonApiEntity person = new PersonApiEntity();
        person.setId(id);
        person.setFirstName(firstName);
        person.setMiddleName(middleName);
        person.setLastName(lastName);
        return person;
    }

    public static CarApiEntity fullCar() {
        return fullCar(
                TestConstants.CAR_ID,
                TestConstants.CAR_YEAR,
                TestConstants.CAR_MAKE,
                TestConstants.CAR_MODEL,
                TestConstants.CAR_TRIM,
                TestConstants.CAR_COLOR
        );
    }

    public static CarApiEntity fullCar(String id, Year year, String make, String model, String trim, String color) {
        CarApiEntity car = new CarApiEntity();
        car.setId(id);
        car.setYear(year);
        car.setMake(make);
        car.setModel(model);
        car.setTrim(trim);
        car.setColor(color);
        return car;
    }

    public static EventApiEntity fullEvent() {
        return fullEvent(
                TestConstants.EVENT_ID,
                TestConstants.EVENT_NAME,
                TestConstants.EVENT_DATE,
                TestConstants.HANDICAP_GROUP_SET_ID,
                TestConstants.COMPETITION_GROUP_SET_ID,
                TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION,
                TestConstants.EVENT_CURRENT
        );
    }

    public static EventApiEntity fullEvent(
            String id,
            String name,
            Date date,
            String handicapGroupSetId,
            String competitionGroupSetId,
            int maxRunsPerRegistration,
            boolean current
    ) {
        EventApiEntity event = new EventApiEntity();
        event.setId(id);
        event.setName(name);
        event.setDate(date);
        event.setHandicapGroupSetId(handicapGroupSetId);
        event.setCompetitionGroupSetId(competitionGroupSetId);
        event.setMaxRunsPerRegistration(maxRunsPerRegistration);
        event.setCurrent(current);
        return event;
    }

    public static RegistrationApiEntity fullRegistration() {
        return fullRegistration(
                TestConstants.REGISTRATION_ID,
                fullPerson(),
                fullCar(),
                TestConstants.HANDICAP_GROUP_ID,
                TestConstants.COMPETITION_GROUP_ID,
                TestConstants.REGISTRATION_NUMBER,
                TestConstants.REGISTRATION_CHECKED_IN
        );
    }

    public static RegistrationApiEntity fullRegistration(
            String id,
            PersonApiEntity person,
            CarApiEntity car,
            String handicapGroupId,
            String competitionGroupId,
            String number,
            boolean checkedIn
    ) {
        RegistrationApiEntity registration = new RegistrationApiEntity();
        registration.setId(id);
        registration.setPerson(person);
        registration.setCar(car);
        registration.setHandicapGroupId(handicapGroupId);
        registration.setCompetitionGroupId(competitionGroupId);
        registration.setNumber(number);
        registration.setCheckedIn(checkedIn);
        return registration;
    }

    public static HandicapGroupApiEntity fullHandicapGroup() {
        return fullHandicapGroup(
                TestConstants.HANDICAP_GROUP_ID,
                TestConstants.HANDICAP_GROUP_NAME,
                TestConstants.HANDICAP_GROUP_FACTOR
        );
    }

    public static HandicapGroupApiEntity fullHandicapGroup(
            String id,
            String name,
            BigDecimal factor
    ) {
        HandicapGroupApiEntity handicapGroup = new HandicapGroupApiEntity();
        handicapGroup.setId(id);
        handicapGroup.setName(name);
        handicapGroup.setFactor(factor);
        return handicapGroup;
    }

    public static HandicapGroupSetApiEntity fullHandicapGroupSet(
            String id,
            String name,
            Set<HandicapGroupApiEntity> handicapGroups
    ) {
        HandicapGroupSetApiEntity handicapGroupSet = new HandicapGroupSetApiEntity();
        handicapGroupSet.setId(id);
        handicapGroupSet.setName(name);
        handicapGroupSet.setHandicapGroups(handicapGroups);
        return handicapGroupSet;
    }

    public static HandicapGroupSetApiEntity fullHandicapGroupSet() {
        return fullHandicapGroupSet(
                TestConstants.HANDICAP_GROUP_SET_ID,
                TestConstants.HANDICAP_GROUP_SET_NAME,
                Sets.newHashSet(fullHandicapGroup())
        );
    }

    public static CompetitionGroupApiEntity fullCompetitionGroup() {
        return fullCompetitionGroup(
                TestConstants.COMPETITION_GROUP_ID,
                TestConstants.COMPETITION_GROUP_NAME,
                TestConstants.COMPETITION_GROUP_FACTOR,
                TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.name(),
                TestConstants.COMPETITION_GROUP_GROUPING
        );
    }

    public static CompetitionGroupApiEntity fullCompetitionGroup(String id, AddCompetitionGroupRequest addRequest) {
        return fullCompetitionGroup(
                id,
                addRequest.getName(),
                addRequest.getFactor(),
                addRequest.getResultTimeType(),
                addRequest.getGrouping()
        );
    }

    public static CompetitionGroupApiEntity fullCompetitionGroup(
            String id,
            String name,
            BigDecimal handicapFactor,
            String resultTimeType,
            boolean grouping
    ) {
        CompetitionGroupApiEntity competitionGroup = new CompetitionGroupApiEntity();
        competitionGroup.setId(id);
        competitionGroup.setName(name);
        competitionGroup.setFactor(handicapFactor);
        competitionGroup.setResultTimeType(resultTimeType);
        competitionGroup.setGrouping(grouping);
        return competitionGroup;
    }

    public static CompetitionGroupSetApiEntity fullCompetitionGroupSet(
            String id,
            String name,
            Set<CompetitionGroupApiEntity> competitionGroups
    ) {
        CompetitionGroupSetApiEntity competitionGroupSet = new CompetitionGroupSetApiEntity();
        competitionGroupSet.setId(id);
        competitionGroupSet.setName(name);
        competitionGroupSet.setCompetitionGroups(competitionGroups);
        return competitionGroupSet;
    }

    public static CompetitionGroupSetApiEntity fullCompetitionGroupSet() {
        return fullCompetitionGroupSet(
                TestConstants.COMPETITION_GROUP_SET_ID,
                TestConstants.COMPETITION_GROUP_SET_NAME,
                Sets.newHashSet(fullCompetitionGroup())
        );
    }

    public static RunApiEntity fullRun() {
        return RunApiEntity.builder()
                .id(TestConstants.RUN_ID)
                .eventId(TestConstants.EVENT_ID)
                .registrationId(TestConstants.REGISTRATION_ID)
                .sequence(TestConstants.RUN_SEQUENCE)
                .timestamp(TestConstants.RUN_TIMESTAMP)
                .rawTime(TestConstants.RUN_RAW_TIME)
                .cones(TestConstants.RUN_CONES)
                .didNotFinish(TestConstants.RUN_DID_NOT_FINISH)
                .disqualified(TestConstants.RUN_DISQUALIFIED)
                .rerun(TestConstants.RUN_RERUN)
                .competitive(TestConstants.RUN_COMPETITIVE)
                .build();
    }
}

File: service/src/test/java/org/coner/core/util/ApiEntityTestUtils.kt
package org.coner.core.util

import com.google.common.collect.Sets
import org.coner.core.api.entity.*
import java.math.BigDecimal
import java.time.Instant
import java.time.Year
import java.util.*

object ApiEntityTestUtils {

    @JvmOverloads
    @JvmStatic
    fun fullPerson(id: String = TestConstants.PERSON_ID, firstName: String = TestConstants.PERSON_FIRST_NAME, middleName: String = TestConstants.PERSON_MIDDLE_NAME, lastName: String = TestConstants.PERSON_LAST_NAME): PersonApiEntity {
        val person = PersonApiEntity()
        person.id = id
        person.firstName = firstName
        person.middleName = middleName
        person.lastName = lastName
        return person
    }

    @JvmOverloads
    @JvmStatic
    fun fullCar(id: String = TestConstants.CAR_ID, year: Year = TestConstants.CAR_YEAR, make: String = TestConstants.CAR_MAKE, model: String = TestConstants.CAR_MODEL, trim: String = TestConstants.CAR_TRIM, color: String = TestConstants.CAR_COLOR): CarApiEntity {
        val car = CarApiEntity()
        car.id = id
        car.year = year
        car.make = make
        car.model = model
        car.trim = trim
        car.color = color
        return car
    }

    @JvmOverloads
    @JvmStatic
    fun fullEvent(
            id: String = TestConstants.EVENT_ID,
            name: String = TestConstants.EVENT_NAME,
            date: Date = TestConstants.EVENT_DATE,
            handicapGroupSetId: String = TestConstants.HANDICAP_GROUP_SET_ID,
            competitionGroupSetId: String = TestConstants.COMPETITION_GROUP_SET_ID,
            maxRunsPerRegistration: Int = TestConstants.EVENT_MAX_RUNS_PER_REGISTRATION,
            current: Boolean = TestConstants.EVENT_CURRENT
    ): EventApiEntity {
        val event = EventApiEntity()
        event.id = id
        event.name = name
        event.date = date
        event.handicapGroupSetId = handicapGroupSetId
        event.competitionGroupSetId = competitionGroupSetId
        event.maxRunsPerRegistration = maxRunsPerRegistration
        event.isCurrent = current
        return event
    }

    @JvmOverloads
    @JvmStatic
    fun fullRegistration(
            id: String = TestConstants.REGISTRATION_ID,
            person: PersonApiEntity = fullPerson(),
            car: CarApiEntity = fullCar(),
            handicapGroupId: String = TestConstants.HANDICAP_GROUP_ID,
            competitionGroupId: String = TestConstants.COMPETITION_GROUP_ID,
            number: String = TestConstants.REGISTRATION_NUMBER,
            checkedIn: Boolean = TestConstants.REGISTRATION_CHECKED_IN
    ): RegistrationApiEntity {
        val registration = RegistrationApiEntity()
        registration.id = id
        registration.person = person
        registration.car = car
        registration.handicapGroupId = handicapGroupId
        registration.competitionGroupId = competitionGroupId
        registration.number = number
        registration.isCheckedIn = checkedIn
        return registration
    }

    @JvmOverloads
    @JvmStatic
    fun fullHandicapGroup(
            id: String = TestConstants.HANDICAP_GROUP_ID,
            name: String = TestConstants.HANDICAP_GROUP_NAME,
            factor: BigDecimal = TestConstants.HANDICAP_GROUP_FACTOR
    ): HandicapGroupApiEntity {
        val handicapGroup = HandicapGroupApiEntity()
        handicapGroup.id = id
        handicapGroup.name = name
        handicapGroup.factor = factor
        return handicapGroup
    }

    @JvmOverloads
    @JvmStatic
    fun fullHandicapGroupSet(
            id: String = TestConstants.HANDICAP_GROUP_SET_ID,
            name: String = TestConstants.HANDICAP_GROUP_SET_NAME,
            handicapGroups: Set<HandicapGroupApiEntity> = Sets.newHashSet(fullHandicapGroup())
    ): HandicapGroupSetApiEntity {
        val handicapGroupSet = HandicapGroupSetApiEntity()
        handicapGroupSet.id = id
        handicapGroupSet.name = name
        handicapGroupSet.handicapGroups = handicapGroups
        return handicapGroupSet
    }

    @JvmOverloads
    @JvmStatic
    fun fullCompetitionGroup(
            id: String = TestConstants.COMPETITION_GROUP_ID,
            name: String = TestConstants.COMPETITION_GROUP_NAME,
            handicapFactor: BigDecimal = TestConstants.COMPETITION_GROUP_FACTOR,
            resultTimeType: String = TestConstants.COMPETITION_GROUP_RESULT_TIME_TYPE.name,
            grouping: Boolean = TestConstants.COMPETITION_GROUP_GROUPING
    ): CompetitionGroupApiEntity {
        val competitionGroup = CompetitionGroupApiEntity()
        competitionGroup.id = id
        competitionGroup.name = name
        competitionGroup.factor = handicapFactor
        competitionGroup.resultTimeType = resultTimeType
        competitionGroup.isGrouping = grouping
        return competitionGroup
    }

    @JvmOverloads
    @JvmStatic
    fun fullCompetitionGroupSet(
            id: String = TestConstants.COMPETITION_GROUP_SET_ID,
            name: String = TestConstants.COMPETITION_GROUP_SET_NAME,
            competitionGroups: Set<CompetitionGroupApiEntity> = Sets.newHashSet(fullCompetitionGroup())
    ): CompetitionGroupSetApiEntity {
        val competitionGroupSet = CompetitionGroupSetApiEntity()
        competitionGroupSet.id = id
        competitionGroupSet.name = name
        competitionGroupSet.competitionGroups = competitionGroups
        return competitionGroupSet
    }

    @JvmOverloads
    @JvmStatic
    fun fullRun(
           id: String = TestConstants.RUN_ID,
           eventId: String = TestConstants.EVENT_ID,
           registrationId: String = TestConstants.REGISTRATION_ID,
           sequence: Int = TestConstants.RUN_SEQUENCE,
           timestamp: Instant = TestConstants.RUN_TIMESTAMP,
           rawTime: BigDecimal = TestConstants.RUN_RAW_TIME,
           cones: Int = TestConstants.RUN_CONES,
           didNotFinish: Boolean = TestConstants.RUN_DID_NOT_FINISH,
           disqualified: Boolean = TestConstants.RUN_DISQUALIFIED,
           rerun: Boolean = TestConstants.RUN_RERUN,
           competitive: Boolean = TestConstants.RUN_COMPETITIVE
    ) = RunApiEntity(
                id = id,
                eventId = eventId,
                registrationId = registrationId,
                sequence = sequence,
                timestamp = timestamp,
                rawTime = rawTime,
                cones = cones,
                didNotFinish = didNotFinish,
                disqualified = disqualified,
                rerun = rerun,
                competitive = competitive
        )
}
